sweep/sweeper.go

```diff
@@ -1326,14 +1593,81 @@ func (s *UtxoSweeper) handleInputSpent(spend *chainntnfs.SpendDetail) {
 	}
 }
 
-// handleSweep is called when the ticker fires. It will create clusters and
-// attempt to create and publish the sweeping transactions.
-func (s *UtxoSweeper) handleSweep() {
+// markInputFailed marks the given input as failed and won't be retried. It
+// will also notify all the subscribers of this input.
+func (s *UtxoSweeper) markInputFailed(pi *pendingInput, err error) {
+	log.Errorf("Failed to sweep input: %v, error: %v", pi, err)
+
+	pi.state = StateFailed
+
+	// Remove all other inputs in this exclusive group.
+	if pi.params.ExclusiveGroup != nil {
+		s.removeExclusiveGroup(*pi.params.ExclusiveGroup)
+	}
+
+	s.signalResult(pi, Result{Err: err})
+}
+
+// updateSweeperInputs updates the sweeper's internal state and returns a map
+// of inputs to be swept. It will remove the inputs that are in final states,
+// and returns a map of inputs that have either StateInit or
+// StatePublishFailed.
+func (s *UtxoSweeper) updateSweeperInputs() pendingInputs {
+	// Create a map of inputs to be swept.
+	inputs := make(pendingInputs)
+
+	// Iterate the pending inputs and update the sweeper's state.
+	//
+	// TODO(yy): sweeper is made to communicate via go channels, so no
+	// locks are needed to access the map. However, it'd be safer if we
+	// turn this pendingInputs into a SyncMap in case we wanna add
+	// concurrent access to the map in the future.
```

Comment:
I support the SyncMap thing but I would say we should be moving towards a structure of only ever managing concurrency with go channels for systems that have goroutines (a `Start` method).

Reply:
Discussed offline - totally agree, we should make sure we are communicating via channels, and the `CreateSweepTx` method is removed in the final PR so we now have only `SweepInput` as the entrypoint.